### ISSUE TITLE
`collapsible_if` should have some sort of maximum complexity

### ISSUE THREAD

**Issue by Nemo157** [Manishearth/rust-clippy#789]
Not sure if cyclomatic complexity of the checks or just purely span size should be taken into account, but with the following snippet of code (taken from [here](https://git.nemo157.com/grarr/blob/05dc0d51fa6659ee5d01cf318524909a6ebc0df8/src/handler/git_smart_http/upload_pack.rs#L124)) I get `collapsible_if` warnings that don't really make sense to implement

``` rust
// if it is an ancestor of a ref
if try!(graph_ancestor_of_any(&context.repository, id, context.refs.iter().cloned())) {
  // and is not an ancestor of a common
  if !try!(graph_ancestor_of_any(&context.repository, id, common.iter().cloned())) {
    // add it to common
    common.insert(id);
  }
}
```

warning message:

```
src/handler/git_smart_http/upload_pack.rs:124:5: 130:6 warning: this if statement can be collapsed, #[warn(collapsible_if)] on by default
src/handler/git_smart_http/upload_pack.rs:124     if try!(graph_ancestor_of_any(&context.repository, id, context.refs.iter().cloned())) {
src/handler/git_smart_http/upload_pack.rs:125       // and is not an ancestor of a common
src/handler/git_smart_http/upload_pack.rs:126       if !try!(graph_ancestor_of_any(&context.repository, id, common.iter().cloned())) {
src/handler/git_smart_http/upload_pack.rs:127         // add it to common
src/handler/git_smart_http/upload_pack.rs:128         common.insert(id);
src/handler/git_smart_http/upload_pack.rs:129       }
                                              ...
src/handler/git_smart_http/upload_pack.rs:124:5: 130:6 help: try
src/handler/git_smart_http/upload_pack.rs:        if (try!(graph_ancestor_of_any(&context.repository, id, context.refs.iter().cloned()))) && (!try!(graph_ancestor_of_any(&context.repository, id, common.iter().cloned()))) {
src/handler/git_smart_http/upload_pack.rs:      // add it to common
src/handler/git_smart_http/upload_pack.rs:      common.insert(id);
src/handler/git_smart_http/upload_pack.rs:    }
src/handler/git_smart_http/upload_pack.rs:124:5: 130:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#collapsible_if
```

If I were to follow the suggestion it would result in:

``` rust
// if it is an ancestor of a ref and is not an ancestor of a common
if try!(graph_ancestor_of_any(&context.repository, id, context.refs.iter().cloned())) && !try!(graph_ancestor_of_any(&context.repository, id, common.iter().cloned())) {
  // add it to common
  common.insert(id);
}
```

which is obviously far too long to actually use (and line breaks in conditional checks are horrible). The only "correct" solution to this warning that I can think of without pulling these tests out to helper functions would be something like:

``` rust
let refs_iter = context.refs.iter().cloned();
let common_iter = common.iter().cloned();
let ancestor_of_ref = try!(graph_ancestor_of_any(&context.repository, id, refs_iter));
let ancestor_of_common = || graph_ancestor_of_any(&context.repository, id, common_iter);
// if it is an ancestor of a ref and is not an ancestor of a common
if ancestor_of_ref && !try!(ancestor_of_common()) {
  // add it to common
  common.insert(id);
}
```

which results in needing to pull the iter creation out to stay under 100 characters and using a closure to retain the short-circuiting nature of `&&` with these relatively expensive tests.

**Comment by mcarton** [comment]
Clippy does try to format the suggestion. You can just split the the condition on two lines.
Or alternatively
```rust
let is_ancestor_of = |e| graph_ancestor_of_any(&context.repository, id, e.iter().cloned());
if try!(is_ancestor_of(context.refs)) && try!(is_ancestor_of(common)) {
    common.insert(id);
}
```
Or just `#[allow]` the lint.
I’m not sure the additional complexity is worth it. Especially since most lints that suggest new code might be affected by bad formatting.

**Comment by Nemo157** [comment]
I probably should have been a bit less strong in my initial post, I mean this more as a potential suggestion (`collapsible_if` _could_ have some sort of maximum complexity). I'm not entirely certain that this would be a good idea, the lint popping up here has made me at least re-evaluate whether these conditions should be pulled out to separate variables/helper functions :smile:.

Thinking about it more the main reason I want to keep these as two separate conditions are they have a lot of _semantic_ weight, combining them into a single check feels like it would not give them the individual importance they should have when re-reading this code in the future. Unfortunately that doesn't seem like something that clippy will be able to detect until it gains near-sentience :disappointed:. I've used `#[allow]` for now and it's likely that the warning will just vanish once I get round to refactoring this in the future anyway.

**Comment by mcarton** [comment]
> the lint popping up here has made me at least re-evaluate whether these conditions should be pulled out to separate variables/helper functions :smile:. […] I've used #[allow] for now and it's likely that the warning will just vanish once I get round to refactoring this in the future anyway.

There you go, the lint is good as bad as it is today :smile:.

> Unfortunately that doesn't seem like something that clippy will be able to detect until it gains near-sentience :disappointed:.

This is something we plan to implement for a future release :stuck_out_tongue_winking_eye:.

**Comment by oli-obk** [comment]
Can we close this?

**Comment by Nemo157** [comment]
:+1:
